CRI: Add Status into CRI.

Random-Liu opened the pull request:
For https://github.com/kubernetes/kubernetes/issues/35701.
Fixes https://github.com/kubernetes/kubernetes/issues/35701.

This PR added a `Status` call in CRI, and the `RuntimeStatus` is defined as following:

``` protobuf
message RuntimeCondition {
    // Type of runtime condition.
    optional string type = 1;
    // Status of the condition, one of true/false.
    optional bool status = 2;
    // Brief reason for the condition's last transition.
    optional string reason = 3;
    // Human readable message indicating details about last transition.
    optional string message = 4;
}

message RuntimeStatus {
    // Conditions is an array of current observed runtime conditions.
    repeated RuntimeCondition conditions = 1;
}
```

Currently, only `conditions` is included in `RuntimeStatus`, and the definition is almost the same with `NodeCondition` and `PodCondition` in K8s api.

@yujuhong @feiskyer @bprashanth If this makes sense, I'll send a follow up PR to let dockershim return `RuntimeStatus` and let kubelet make use of it.
@yifan-gu @euank Does this make sense to rkt?
/cc @kubernetes/sig-node

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/35839)

<!-- Reviewable:end -->

Random-Liu committed:
Add Status into CRI.

euank commented:
I don't think there's an issue with this from the rkt perspective. The same hostNetwork thing shhould behave pretty much identically here.

I do question the original choice of hostNetworkOnly being a condition though... thinking about it more, being able to reject a pod with a status like "networkNotReady" or "runtimeNotReady" as a reason would allow the kubelet to know to retry things in those cases and might be a more reliable way to handle these things.

bprashanth commented:
rkt can just piggyback on kubenet's status(), which is a synthetic api to begin with (meaning it's not actually a part of CNI). 

Status does 2 things:     

1. checks if the cni binary exists at pluginDir
2. checks if the cni conf has been initialized with podCIDR

If either of those is false, the node remains notready. 
If --runtime=rkt --plugin=!kubenet, the rkt cri will have to implement a status() that effectively does the above.

bprashanth commented:
@kubernetes/sig-network

bprashanth commented on pkg/kubelet/dockershim/remote/docker_service.go:
```diff
@@ -49,6 +49,10 @@ func (d *dockerService) Version(ctx context.Context, r *runtimeApi.VersionReques
 }
 
+func (d *dockerService) Status(ctx context.Context, r *runtimeApi.StatusRequest) (*runtimeApi.StatusResponse, error) {
+	return nil, fmt.Errorf("not implemented")
```
Please add godoc. how are we planning to integrate this into the `defaultNodeStatusFuncs`? I'm guessing it needs to make its way into `runtimeState.networkErrors`?

yujuhong replied:
Yes, we probably need to change [`syncNetworkStatus`](https://github.com/kubernetes/kubernetes/blob/2941aea88e80e94f2698241a4ac9925145664470/pkg/kubelet/kubelet_network.go#L193) to pipe this information.

bprashanth replied:
Shouldn't the kubelet parses out network/runtime status directly from the response of this rpc in syncRuntimeStatus?

bprashanth commented on pkg/kubelet/api/v1alpha1/runtime/api.pb.go:
```diff
@@ -2603,6 +2607,109 @@ func (m *UpdateRuntimeConfigResponse) String() string            { return proto.
+// properly, including:
+//   * RuntimeReady: RuntimeReady means the runtime is up and ready to accept
+//   basic containers.
+//   * NetworkReady: NetworkReady means the runtime network is up and ready to
```
so if any required condition is unmet the node remains notready (please elaborate comment)? 
we will probably have to conditionally start hostnetwork pods if NetworkReady == false but everything else checks out.

Random-Liu replied:
ACK

yujuhong reviewed (changes requested):
Need to update the internal interface in `pkg/kubelet/api/services.go` and add an implementation stub in `dockershim/docker_service.go`

Random-Liu commented:
Will finish the following part today.

feiskyer commented:
> Need to update the internal interface in pkg/kubelet/api/services.go and add an implementation stub in dockershim/docker_service.go

Besides, also add `Status()` calling at `pkg/kubelet/remote` and `pkg/kubelet/api/testing`.

Random-Liu committed:
Add Status implementation.

Random-Liu commented:
Addressed comments and implement the initial version of `Status`.

I'll send another PR to expose RuntimeReady and NetworkReady from runtime interface https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/container/runtime.go#L73.

Random-Liu commented:
@k8s-bot cri test this.

Random-Liu commented on pkg/kubelet/api/testing/fake_runtime_service.go:
```diff
@@ -50,10 +50,21 @@ type FakeRuntimeService struct {
 
+// SetFakeRuntime is a simple function to set arbitrary field in the fake runtime
+// with lock.
+func (r *FakeRuntimeService) SetFakeRuntime(set func(*FakeRuntimeService)) {
```
Notice that this is not used, but will be used in the future unit test.

Random-Liu commented:
@k8s-bot cri test this.

Random-Liu committed:
Add RuntimeStatus in container/runtime.go

Random-Liu committed:
Update bazel.

feiskyer commented on pkg/kubelet/api/testing/fake_runtime_service.go:
```diff
@@ -50,10 +50,21 @@ type FakeRuntimeService struct {
 
 	Called []string
 
+	Errors     map[string]error
```
what is this for?

Random-Liu replied:
This is for injecting errors.

feiskyer commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -261,13 +261,23 @@ func (m *kubeGenericRuntimeManager) APIVersion() (kubecontainer.Version, error)
+	if runtimeReady == nil || !runtimeReady.GetStatus() {
+		return fmt.Errorf("runtime not ready: reason: %q, message: %q",
+			runtimeReady.GetReason(), runtimeReady.GetMessage())
 	}
```
nit: do we need to report node events if there are other not ready conditions?

Random-Liu replied:
The code is temporary, I'll send another PR/commit soon to really use the conditions. I'll deal with then.

feiskyer replied:
LGTM

Random-Liu committed:
Populate NetworkReady Status.

Random-Liu commented:
@k8s-bot cri test this.

Random-Liu commented:
PR is ready for review.

Random-Liu commented:
@k8s-bot cri test this.

feiskyer commented:
LGTM. needs @yujuhong take another look, and confirm review changes.

yujuhong commented on pkg/kubelet/kubelet.go:
```diff
@@ -1899,10 +1893,40 @@ func (kl *Kubelet) PLEGHealthCheck() (bool, error) {
 	}
+	// Only check specific conditions when runtime integration type is cri,
+	// because the old integration doesn't populate any runtime condition.
+	if kl.kubeletConfiguration.ExperimentalRuntimeIntegrationType == "cri" {
```
Never mind. I decided to send in #36201 to enforce that all integration using CRI should specify --experimental-runtime-integration-type=cri for now to simplify the code.

yujuhong reviewed (approved):
LGTM.

Random-Liu commented:
@yujuhong Add the `networkPlugin=nil` back.

Random-Liu commented:
Just rebase, apply LGTM again.
